### PR TITLE
Upgrade Mediawiki's version to 1.34

### PR DESCRIPTION
**Pull request by icyflame**
## Summary

The nightly Travis CI build has been failing for the past few days as ArticleFeedback doesn't support Mediawiki 1.33. Mediawiki 1.34 is now stable.

This PR updates the Mediawiki version to 1.34.

Every commit explains the reasons for the change. Here's a summary:

- Scribunto is shipped with mediawiki-core now, we don't need any configuration on our side
- PHP 7.1 is no longer supported (End-of-Life starting Dec 2019) => I updated the versions to PHP 7.3
- Debian Jessia is oldstable version, I upgraded it to the latest stable version: Debian Buster
- The RecentPages extension uses the PHP constant `DB_SLAVE` but this constant was never defined anywhere. PHP 7.2 starts raising a warning if we refer to an undefined constant. PHP 7.3 throws an error.
    - This value is passed as the first argument to `wfGetDB` by RecentPages. It _must_ be an integer.

image

Here's a summary of the process I went through during the upgrade:

<details>
<summary>Mediawiki 1.33 => 1.34 Upgrade summary</summary>
<pre>
This periodic failing build on Travis CI got to me

https://travis-ci.org/metakgp/metakgp-wiki/builds/656667168#L2620

I am making a PR to upgrade to Mediawiki 1.34.

icyflame  1 hour ago

- Scribunto is part of mediawiki core now
- We don't need to pull the master branch of ArticleFeedbackV5 anymore
- PHP 7 and 7.1 aren't supported  I upgraded to PHP 7.3 with Debian Buster (We
  were using Debian Jessie, which is two releases ago)
    - 7.1 => 7.3 had a couple of breaking changes: count(null) is an error now.
      Hopefully none of the extensions has some code like that

icyflame  1 hour ago

On the bright side though, 7.3 is supposed to have some major improvements
related to RAM usage. So, maybe our memory usage will go down from 90 => 80%?

icyflame  1 hour ago

Oh, Node 10 on Parsoid and Python 2 on Backup  We have to fix that.  I will
create a GitHub issue for each of them.

icyflame  1 hour ago

Issues created//github.com/metakgp/metakgp-wiki/issues/76
https://github.com/metakgp/metakgp-wiki/issues/77

icyflame  1 hour ago

I didn't remove all the Scribunto configuration  Removed and re-building now.

icyflame  39 minutes ago

Alright, we have our first customer: RecentPages has some strange code?

> 2020/02/29 15 PHP Warning:  Use of undefined constant DB_SLAVE - assumed
> 'DB_SLAVE' (this will throw an Error in a future version of PHP) in
> /srv/mediawiki/extensions/RecentPages/RecentPages.php on line 279" while
> reading response header from upstream, client wiki.metakgp.org, request9000",
> host8080"

icyflame  35 minutes ago

We are fetching the PHP file from
GitHubhttps://github.com/Inclumedia/RecentPages/blob/b7d863d2b862e70ebde0635e7845b4cba0d4db7e/RecentPages.php#L279

Code looks fine, maybe I just need to define DB_SLAVE?

icyflame  30 minutes ago

> If you use an undefined constant, PHP assumes that you mean the name of the
> constant itself, just as if you called it as a string (CONSTANT vs
> "CONSTANT"). This fallback is deprecated as of PHP 7.2.0, and an error of
> level E_WARNING is issued when it happens (previously, an error of level
> E_NOTICE has been issued instead.) See also the manual entry on why $foo[bar]
> is wrong (unless you first define() bar as a constant). This does not apply to
> (fully) qualified constants, which will raise a fatal error if undefined. If
> you simply want to check if a constant is set, use the defined() function.

-- https://www.php.net/manual/en/language.constants.syntax.php

So, we just need to explicitly define the DB_SLAVE constant and set it to the
value "DB_SLAVE" (edited)

icyflame  30 minutes ago

Done, re-building now;

icyflame  29 minutes ago

Didn't add a semicolon. I need a stricter PHP linter in my editor.  Fixed and
re-building

icyflame  23 minutes ago

All the pages except the main page work fine. Here we go.  (edited)

icyflame  15 minutes ago

> [4ffcc58250ce869ee5be1905] /w/Main_Page UnexpectedValueException from line 462
> of /srv/mediawiki/includes/libs/rdbms/loadbalancer/LoadBalancer.php: Invalid
> server index index #DB_SLAVE

Still looks like a RecentPages issue, the value I gave this constant was
incorrect?  I wonder what was the constant's value in the previous PHP version.
Perhaps, I should find that.

icyflame  13 minutes ago

https://doc.wikimedia.org/mediawiki-core/master/php/GlobalFunctions_8php.html#a46345f25555fd93d272119932fbfc18f

Ah, it should be an integer.

icyflame  8 minutes ago

AHA! WORKED! (edited)
</pre>
</details>

### CODE REVIEW

**Review by thealphadollar — APPROVED** [review]

**Comment by thealphadollar** [comment]
I'll try to merge this tomorrow :)

**Comment by icyflame** [comment]
@thealphadollar Hi, I would like to release this PR. Also, I would like to wait for Vikrant's review.

**Comment by thealphadollar** [comment]
Sounds good to me.

File: mediawiki/LocalSettings.php
Comment (amrav) [inline mediawiki/LocalSettings.php:157]: The extension page for RecentChanges says it's archived and suggests using https://www.mediawiki.org/wiki/Extension:SimpleChanges. Not blocking, but seems like a simple change and if it does the same job, would be better to use that instead of trying to support a deprecated extension ourselves :)

(Great job debugging this though!)

File: mediawiki/LocalSettings.php
Comment (amrav) [inline mediawiki/LocalSettings.php:170]: Do we still need these lines of configuration? Would be nice to cleanup if not.

File: mediawiki/install_extensions.sh
Comment (amrav) [inline mediawiki/install_extensions.sh:22]: Thanks for cleanup! Does AF work now, or does it still have the same issues?

File: mediawiki/Dockerfile
Comment (amrav) [inline mediawiki/Dockerfile:1]: Wdyt about leaving this as `php:7-fpm-buster` so we get the latest minor versions?

**Review by amrav — APPROVED** [review]
Generally looks great, some non-blocking comments :)

OOC did you notice any memory reduction after switching to PHP 7.3?

File: mediawiki/install_extensions.sh
Comment (icyflame) [reply mediawiki/install_extensions.sh:22]: It still has the same issues! I am unable to `Resolve feedback` on the local deployment. 😢 

But at least now it reproduces locally! With 1.33, it would not reproduce locally. So, I can turn on debug and fix it hopefully.

File: mediawiki/LocalSettings.php
Comment (icyflame) [reply mediawiki/LocalSettings.php:170]: I thought about these two lines and the `chmod a+x` line in `install_extensions`, I need to investigate some more to find out what the default mode of that file is + what the default values for these two variables are.

I will do this investigation later and create a separate PR! I agree, it would be nice to clean-up 😅

File: mediawiki/LocalSettings.php
Comment (icyflame) [reply mediawiki/LocalSettings.php:157]: Oh, I see! We should swap out this for RecentChanges then.

File: mediawiki/Dockerfile
Comment (icyflame) [reply mediawiki/Dockerfile:1]: Oh, actually, we can't do that because Mediawiki 1.34 drops support from PHP 7.1 😢 

> Some specific notes for MediaWiki 1.34 upgrades are below:
> * MediaWiki now requires PHP 7.2.9 or above.
>
-- https://gerrit.wikimedia.org/r/plugins/gitiles/mediawiki/core/+/REL1_34/RELEASE-NOTES-1.34#90

**Comment by icyflame** [comment]
🗒 I created #79, #80, and #81 to make some changes once this PR is merged and deployed.

I will deploy this PR tonight.

**Pull request merged**
